Effects 1.0.0 + Todo Sample

semanticer opened the pull request:
## Summary
This is a ~~very incomplete first draft~~ of Effects 1.0.0 implementation. ~~The implementation is extremely naive and without cancellation functionality implemented.~~ I also implemented the Todo Sample to better visualize how everything firsts together. 

### Effect implementation
For this initial draft PR, I simply changed the suspended function to a function returning flow:
```kotlin
fun interface Effect<out Action> {
    fun actions(): Flow<Action>

    companion object
}

object NoEffect : Effect<Nothing> {
    override fun actions(): Flow<Nothing> = emptyFlow()
}
```

But there are other ways we can implement this. We could do something similar to TCA using sealed class structure. This is how the TCA's version looks like: 

```swift
public struct Effect<Action> {
  @usableFromInline
  enum Operation {
    case none
    case publisher(AnyPublisher<Action, Never>)
    case run(TaskPriority? = nil, @Sendable (_ send: Send<Action>) async -> Void)
  }

  @usableFromInline
  let operation: Operation

  @usableFromInline
  init(operation: Operation) {
    self.operation = operation
  }
}
```

Then there is [kotlin-composable-architecture's](https://github.com/wearemakery/kotlin-composable-architecture) approach: 

```kotlin
class Effect<Output>(internal var flow: Flow<Output>) {

    companion object {
        operator fun <Output> invoke(
            block: suspend FlowCollector<Output>.() -> Unit
        ): Effect<Output> = Effect(flow(block))

        fun <Output> none() = Effect<Output>(emptyFlow())
    }

    fun <T> map(transform: (Output) -> T): Effect<T> = Effect(flow.map { transform(it) })

    fun concatenate(vararg effects: Effect<Output>) {
        flow = flowOf(flow, *effects.map { it.flow }.toTypedArray()).flattenConcat()
    }

    fun merge(vararg effects: Effect<Output>) {
        flow = flowOf(flow, *effects.map { it.flow }.toTypedArray()).flattenMerge()
    }

    suspend fun sink(): List<Output> {
        val outputs = mutableListOf<Output>()
        flow.toList(outputs)
        return outputs
    }
}
```

### Effect creation API
Let me start with a comparison from the Sample Todo app

This PR:

```kotlin
is TodosAction.Todo -> if (action.action is TodoAction.IsCompleteChanged) {
    state.withEffect {
        suspended {
            delay(1000)
            TodosAction.SortCompletedTodos
        }.cancellable(key = "sort", cancelInFlight = true)
    }
} else {
    state.withoutEffect()
}
```

[kotlin-composable-architecture's](https://github.com/wearemakery/kotlin-composable-architecture) implementation:

```kotlin
is AppAction.Todo -> {
    if (action.action is TodoAction.CheckBoxToggled) {
        state
            .withEffect<AppState, AppAction> {
                delay(1000L)
                emit(AppAction.SortCompletedTodos)
            }
            .cancellable("TodoCompletionId", cancelInFlight = true)
    } else {
        state.withNoEffect()
    }
}
```
☝️ Note, that `Result` needs to be created twice here, first with the original effect and then again with it's cancellable version.

TCA's syntax can't be achieved in our context but I'll add it here anyway for reference:

```swift
  case .todo(id: _, action: .binding(\.$isComplete)):
    return .run { send in
      try await self.clock.sleep(for: .seconds(1))
      await send(.sortCompletedTodos, animation: .default)
    }
    .cancellable(id: CancelID.todoCompletion, cancelInFlight: true)
```

### Cancelation implementation
As noted earlier, I haven't tackled the implementation of cancellation logic just yet. I'm currently determining the optimal approach. However, if you're adept with coroutine intricacies and feel compelled to take a crack at it, feel free to do so.

For reference, here is [TCA's implementation](https://github.com/pointfreeco/swift-composable-architecture/blob/main/Sources/ComposableArchitecture/Effects/Cancellation.swift) and here is [kotlin-composable-architecture's](https://github.com/wearemakery/kotlin-composable-architecture/blob/master/composable-architecture/src/main/kotlin/composablearchitecture/Effect%2BCancellation.kt)


Missing stuff: 
- [ ] Cancellation logic
- [ ] More comprehensive tests
- [ ] Finnal API for creating effects 

## Relationships

Closes #62 
Closes #63 

## Tests
✅ 

## Review pointers
This has to work, but we also need to come up with a reasonable API

daividssilverio commented on komposable-architecture/src/test/java/com/toggl/komposable/reducer/CompositeReducerTests.kt:
```diff
@@ -33,9 +34,9 @@ class CompositeReducerTests {
+        combinedReducer.testReduce(originalState, TestAction.DoNothingAction) { state, effect ->
             state shouldBe stateFromSecond
-            effects shouldBe effectsFromFirst + effectsFromSecond
+            effect shouldBe effectsFromFirst.merge(effectsFromSecond)
```
Hey, merge doesn't preserve the order of the elements. It looks like the test shouldn't use the `merge` extension directly if you'd like to test the sequentiality of the returned flow.

semanticer replied:
Yeah, this was some mindless test fix that didn't make sense. I'll remove the effects from the test for now.

semanticer commented:
👆 First try at cancellation functionality, inspired by @daividssilverio and some mutex ideas from [kotlin-composable-architecture's](https://github.com/wearemakery/kotlin-composable-architecture).

Next, I need to add cancellation tests so that we know what's actually going on.

thuutin commented on komposable-architecture/src/main/java/com/toggl/komposable/extensions/EffectExtensions.kt:
```diff
@@ -1,62 +1,86 @@
+        Effect { flowOf(*actions) }
+
+    fun <Action> suspended(func: suspend () -> Action): Effect<Action> =
+        Effect { flow { emit(func()) } }
```
```suggestion
        Effect { flow { func.asFlow() } }
```

semanticer commented:
@thuutin @heytherewill @daividssilverio 

I added a bunch of tests and changed some API again. Cancellation seems to be working, but I'm sure there might be some edge cases. 

Effect is now defined as: 

```kotlin
/**
 * Effects are returned by reducers when they wish to produce a side effect.
 * This can be anything from cpu/io bound operations to changes that simply affect the UI.
 * @see Reducer.reduce
 */
fun interface Effect<out Action> {

    /**
     * Executes the effect. This operation can produce side effects, and it's the
     * responsibility of the class implementing this interface to change threads
     * to prevent blocking the UI when needed.
     * @return An action that will be sent again for further processing
     * @see Store.send
     */
    operator fun invoke(): Flow<Action>

    companion object {
        fun none(): Effect<Nothing> = NoEffect
    }
}

object NoEffect : Effect<Nothing> {
    override fun invoke(): Flow<Nothing> = emptyFlow()
}
```

And they can be created in various ways: 

```kotlin
// simple effects that will return action(s)
state.withEffect(TodosAction.SortCompletedTodos)
// or with flow    
state.withEffect(flowOf(TodosAction.SortCompletedTodos))
// suspended function
state.withSuspendEffect {
    delay(1000)
    TodosAction.SortCompletedTodos
}
// callbackFlow
state.withCallbackEffect {
    delay(1000)
    send(TodosAction.SortCompletedTodos)
    delay(1000)
    send(TodosAction.SortCompletedTodos)
}
// your custom effect instance
state.withEffect(MyEffect())
  
// any of these effects can be easily made cancellable 
state.withSuspendEffect(id = "sort", cancelInFlight = true) {
    delay(1000)
    TodosAction.SortCompletedTodos
}

```

I hope we can further improve the tests after we implement https://github.com/toggl/komposable-architecture/issues/54

daividssilverio commented on komposable-architecture/src/main/java/com/toggl/komposable/extensions/EffectCancellationExtensions.kt:
```diff
@@ -0,0 +1,51 @@
+ * @param action The ID of the effects to be cancelled.
+ * @return An Effect of Nothing that, when executed, will cancel the specified effects.
+ */
+fun Effect.Companion.cancel(action: Any): Effect<Nothing> =
```
Oh, we could have a `cancelAll` to "stop the world".

semanticer replied:
In which context can this be useful?

daividssilverio replied:
Logouts for example.

semanticer replied:
I see, I'm not against it. Maybe ScopedActions as implemented by @sveltema [here](https://github.com/sveltema/KMPosable-architecture/blob/main/kmposable-core/src/commonMain/kotlin/com/labosu/kmposable/Action.kt) could also be a solution by I think `cancelAll` makes sense too 👍

semanticer replied:
done

semanticer commented:
@daividssilverio @thuutin @heytherewill 
I was also thinking... many of the effects from Toggl's main repo specify the context in which they are happening. 

```kotlin
override suspend fun execute(): AddFavoriteAction = withContext(dispatcherProvider.io) {
        favoriteRepository
            .insertFavorite(favorite)
            .let { AddFavoriteAction.FavoriteSuccessfullyAdded }
    }
```

Should we build this in perhaps?

```kotlin
state.withSuspendEffect(id = "add", cancelInFlight = true, context = dispatcherProvider.io) {
        favoriteRepository
            .insertFavorite(favorite)
            .let { AddFavoriteAction.FavoriteSuccessfullyAdded }
} 
```

daividssilverio commented on samples/todos/src/main/java/com/toggl/komposable/sample/todos/Todos.kt:
```diff
@@ -0,0 +1,80 @@
+}
+
+class TodosReducer : Reducer<TodosState, TodosAction> {
+    override fun reduce(state: TodosState, action: TodosAction): ReduceResult<TodosState, TodosAction> =
```
We could add some "fake download" to each item to simulate the cancellable events and test that we can cancel a bunch of ongoing effects in one go.

semanticer replied:
We should do this in tests, I want to keep this sample as close to https://github.com/pointfreeco/swift-composable-architecture/tree/main/Examples/Todos/Todos as possible.

daividssilverio commented:
> Should we build this in perhaps?

```
state.withSuspendEffect(id = "add", cancelInFlight = true, context = dispatcherProvider.io) {
       favoriteRepository
            .insertFavorite(favorite)
            .let { AddFavoriteAction.FavoriteSuccessfullyAdded }
}
```

Looks like a nice to have. 👍

daividssilverio reviewed (commented):
Looking pretty good. Can we generate a snapshot release or something to try this in development?

heytherewill commented on komposable-architecture/src/main/java/com/toggl/komposable/architecture/Effect.kt:
```diff
@@ -14,5 +21,25 @@ fun interface Effect<out Action> {
      * @see Store.send
      */
-    suspend fun execute(): Action?
+    operator fun invoke(): Flow<Action>
```
I don't really like this change. While in the context of my pet projects doing operator overloading is fine, once you have too much of it you create needless layers of indirection. I don't think calling `effect.execute()` is so bad that you wanna omit the `.execute`, specially since this won't be something that the end users need to do often (likely only in tests, where readability is usually important). 

That said, this is not a hill I'd die on 🤷

semanticer replied:
@heytherewill  Yeah, I don't think the extra syntactic sugar that `invoke()` provides is important; I just didn't know how else to call this tbh :-D 
I just wasn't sure if "execute" correctly describes a function that returns flow where it can be technically "executed" later in some sense 🤷 
Can you think of a better name?

heytherewill replied:
`run`? Not sure if because it already exists in another context it would be confusing. `start` would also be a good call

semanticer replied:
I guess both could work, I like `run` better probably, but as you said, `run` from stdlib makes it a bit confusing. What about just `actions` or `actionFlow`. Idk, naming is so hard, help us @daividssilverio, @thuutin

daividssilverio replied:
I like `start` and `actions`, but could be `flow`, `process`, `results`, `produce`, `enqueue`, `breakTheWorld`

`start` and `process` being my favorites. Just `actions` sounds less than it actually is as we can do IO and other stuff inside effects.

heytherewill commented on komposable-architecture/src/main/java/com/toggl/komposable/extensions/EffectCancellationExtensions.kt:
```diff
@@ -0,0 +1,51 @@
+ * @param action The ID of the effects to be cancelled.
+ * @return An Effect of Nothing that, when executed, will cancel the specified effects.
+ */
+fun Effect.Companion.cancel(action: Any): Effect<Nothing> =
```
How is this managed in the Swift TCA? This being a singleton will make testing super hard. It's mega convenient, but a container for the cancellation ids doesn't sound like a bad idea

semanticer replied:
If I can read Swift correctly, it's [pretty much the same thing](https://github.com/pointfreeco/swift-composable-architecture/blob/8d92762a226c486a437ef5263a0661b9112d99fa/Sources/ComposableArchitecture/Effects/Cancellation.swift#L104), right?

heytherewill replied:
Seems about correct. Let's leave it at that, then 🤷

heytherewill commented on komposable-architecture/src/main/java/com/toggl/komposable/extensions/EffectExtensions.kt:
```diff
@@ -1,62 +1,153 @@
+ */
+@OptIn(ExperimentalTypeInference::class)
+fun <State, Action> State.withProducerEffect(
+    id: Any? = null,
+    cancelInFlight: Boolean = false,
```
This combination of parameters allows for invalid state to be expressed `(null, true)`. This rustles my jimmies. We should probably combine them in a sum type instead:

```kotlin
sealed interface CancellableConfiguration {
    data class AllowMultiple(val id: Any) : CancellableConfiguration
    data class CancelInFlight(val id: Any) : CancellableConfiguration
}
```
Names not final. Then we can do:

```suggestion
    configuration: CancellableConfiguration? = null
```

heytherewill replied:
I like your idea of multiple overloads better!

heytherewill commented on komposable-architecture/src/main/java/com/toggl/komposable/internal/CompositeReducer.kt:
```diff
@@ -1,14 +1,15 @@
+        reducers.fold(ReduceResult(state, NoEffect)) { accResult, reducer ->
             val result = reducer.reduce(accResult.state, action)
-            ReduceResult(result.state, accResult.effects + result.effects)
+            ReduceResult(result.state, accResult.effect mergeWith result.effect)
```
You didn't do operator overloading here, which further enhances my point about not doing it for the effect interface

semanticer replied:
Yeah, I feel like `+` means something slightly different than merge does.

heytherewill commented on komposable-architecture/src/main/java/com/toggl/komposable/internal/MutableStateFlowStore.kt:
```diff
@@ -57,31 +58,30 @@ internal class MutableStateFlowStore<State, Action : Any> private constructor(
-                            exceptionHandler.handleEffectException(e)
-                        }
+                    try {
+                        result.effect().onEach {
+                                action ->
+                            send(listOf(action))
+                        }.launchIn(storeScope)
```
Damn, did spotless really format this abomination? 😂

heytherewill replied:
```suggestion
                        result.effect()
                            .onEach { action -> send(listOf(action)) }
                            .launchIn(storeScope)
```

heytherewill commented on komposable-architecture/src/test/java/com/toggl/komposable/effect/EffectCancellationTests.kt:
```diff
@@ -0,0 +1,177 @@
+            effect2Flow
+        }.cancellable("effect", cancelInFlight = false)
+
+        val cancelEffect = Effect.cancel("effect")
```
If the runner does test paralellization the right way, the fact that this test is sharing an effect id with a different test can cause weird errors. Maybe using unique effect ids in the tests is a good way of showing good practices

heytherewill commented on komposable-architecture/src/test/java/com/toggl/komposable/reducer/forEach/ForEachMapTests.kt:
```diff
@@ -29,15 +28,12 @@ class ForEachMapTests {
             when (action) {
                 is ParentAction.EditText ->
-                    ReduceResult(state.copy(parentText = action.text), noEffect())
+                    state.copy(parentText = action.text).withoutEffect()
```
Chef's Kiss

heytherewill reviewed (changes requested):
Looking really good! For such a large change I have very little to say. Can approve once you answer my comments <3

semanticer commented:
> > Should we build this in perhaps?
> 
> ```
> state.withSuspendEffect(id = "add", cancelInFlight = true, context = dispatcherProvider.io) {
>        favoriteRepository
>             .insertFavorite(favorite)
>             .let { AddFavoriteAction.FavoriteSuccessfullyAdded }
> }
> ```
> 
> Looks like a nice to have. 👍

I decided to postpone this as it may make the API a little too complicated. Let's wait for how the dust settles on the other changes, and then we can add this if there is a need.

daividssilverio commented:
Do you think it would be worth adding something like to make it easier to migrate existing effects? Maybe already deprecated. Or maybe we can just use this internally (we are not 1.0.0 yet anyway). I was trying the migration in the Toggl main repo and I started already feeling the pain.

```
abstract class SuspendEffect<out Action> : Effect<Action> {
    abstract suspend fun execute(): Action?
    
    final override fun invoke(): Flow<Action> = flow {
        execute()?.let { emit(it) }
    }
} 
```

thuutin commented on komposable-architecture/src/main/java/com/toggl/komposable/extensions/EffectCancellationExtensions.kt:
```diff
@@ -0,0 +1,51 @@
+        this().onStart {
+            mutex.withLock {
+                if (cancelInFlight) {
+                    cancellationJobs[id]?.forEach { it.cancel() }
```
I think we can pass an exception to `cancel()` to help with debugging.

thuutin commented on komposable-architecture/src/main/java/com/toggl/komposable/extensions/EffectExtensions.kt:
```diff
@@ -1,62 +1,193 @@
- * @param actions          The actions returned by the effects.
- * @return A list of effects.
+ * @param effect An effect to be part of the ReduceResult.
+ * @param id Optional ID to make the effect cancellable. If null, the effect is not cancellable.
```
Is it possible to pass null to a param of type `Any`?

thuutin commented on komposable-architecture/src/main/java/com/toggl/komposable/extensions/EffectCancellationExtensions.kt:
```diff
@@ -0,0 +1,51 @@
+fun <Action> Effect<Action>.cancellable(id: Any, cancelInFlight: Boolean = false): Effect<Action> =
+    Effect {
+        this().onStart {
+            mutex.withLock {
```
Isn't all the flows from `Effect`s are collected in the same thread (the main dispatcher)? If so, do we need the mutex?

semanticer replied:
Tbh, I'm not an expert here, this was inspired by https://github.com/wearemakery/kotlin-composable-architecture/blob/master/composable-architecture/src/main/kotlin/composablearchitecture/Effect%2BCancellation.kt I used mutex as well just to be sure :-D But my understanding is that effects can indeed run in parallel thus mutex might be necessary. I guess we should try to test this 🤔

semanticer commented:
> Do you think it would be worth adding something like to make it easier to migrate existing effects? Maybe already deprecated. Or maybe we can just use this internally (we are not 1.0.0 yet anyway). I was trying the migration in the Toggl main repo and I started already feeling the pain.
> 
> ```kotlin
> abstract class SuspendEffect<out Action> : Effect<Action> {
>     abstract suspend fun execute(): Action?
>     
>     final override fun invoke(): Flow<Action> = flow {
>         execute()?.let { emit(it) }
>     }
> } 
> ```

I like the idea. Let's make the migration easier.

thuutin commented on komposable-architecture/src/main/java/com/toggl/komposable/architecture/Effect.kt:
```diff
@@ -14,5 +21,25 @@ fun interface Effect<out Action> {
      * @return An action that will be sent again for further processing
      * @see Store.send
      */
-    suspend fun execute(): Action?
```
should we also keep this to ease the migration?
```
interface Effect<out Action> {

    @Deprecated(
        "use invoke and return a [Flow] instead",
        replaceWith = ReplaceWith("invoke()")
    )
    suspend fun execute(): Action? = null

    /**
     * Executes the effect. This operation can produce side effects, and it's the
     * responsibility of the class implementing this interface to change threads
     * to prevent blocking the UI when needed.
     * @return An action that will be sent again for further processing
     * @see Store.send
     */
    fun invoke(): Flow<Action> = flow {
        @Suppress("DEPRECATION")
        execute()?.let { emit(it) }
    }
```

semanticer replied:
This would be cool but we'd lose the `fun` modifier from the interface and break many things. I think we shouldn't look back and just provide `SuspendEffect` as a partial remedy instead of compromising effect's design going forward.

semanticer commented on komposable-architecture/src/main/java/com/toggl/komposable/internal/MutableStateFlowStore.kt:
```diff
@@ -57,31 +58,29 @@ internal class MutableStateFlowStore<State, Action : Any> private constructor(
-                        }
+                    try {
+                        result.effect()
+                            .onEach { action -> send(listOf(action)) }
```
I'm not sure I follow, this is pretty much the same case as subscription implementation right?

semanticer commented on komposable-architecture/src/main/java/com/toggl/komposable/extensions/EffectExtensions.kt:
```diff
@@ -1,62 +1,199 @@
  */
-fun <T, R> Effect<T>.map(mapFn: (T?) -> R?): Effect<R> =
-    MapEffect(this, mapFn)
+fun <T, R> Effect<T>.map(mapFn: (T) -> R): Effect<R> =
```
Thanks, makes perfect sense 👍

daividssilverio commented:
@semanticer the todo sample seems to be broken now 🤔 Not sure why though... But currently we can set items as completed, but it isn't possible to revert it. When I was debugging I also noticed that 

a breakpoint on komposable-architecture/samples/todos/src/main/java/com/toggl/komposable/sample/todos/Todos.kt:L55
was being triggered multiple times... not happening when just logs are used (could just be a debugger behaviour).
```
state.withSuspendEffect(id = "sort", cancelInFlight = true) {
                    delay(1000)
                    TodosAction.SortCompletedTodos
                }
```
